### DISCUSSION TITLE
RDFlib Blank node update query

### DISCUSSION THREAD

**Discussion by alex-randles** [RDFLib/rdflib#1605] [Q&A]
Hi All, 

I am trying to update the object of a triple with a blank node as its subject using RDFlib. I firstly select the blank node in the first function and insert this blank node into the update query in the second function, however, this doesn't provide me with the required output. I can't use the add() method or initBindings as I need to save the SPARQL query executed for the user. 

Sample data 
```
@prefix rr: <http://www.w3.org/ns/r2rml#> .
[ rr:objectMap [ rr:column "age" ;
                 rr:language "dhhdhd"] ].
```

Code 
```
mapping_graph = Graph().parse("valid_mapping.ttl",format="ttl")

# find the blank node for the update query 
def find_om_IRI():
    query = """SELECT ?om
                WHERE {
                  ?om rr:language 'dhhdhd' .
                }
           """
    qres = mapping_graph.query(query)
    for row in qres:
        return row[0]
# insert blank node as subject to update query
def change_language_tag():
    om_IRI = find_om_IRI()
    update_query = """
        PREFIX rr: <http://www.w3.org/ns/r2rml#>
        DELETE DATA{
            _:%s  rr:language 'dhhdhd' .
        }
        """ % (om_IRI)
    processUpdate(mapping_graph, update_query)
    print(update_query)
    print(mapping_graph.serialize(format="ttl").decode("utf-8"))
    return update_query


change_language_tag()
```

This however returns the following output. Leaving the graph unchanged. 
```
@prefix rr: <http://www.w3.org/ns/r2rml#> .
[ rr:objectMap [ rr:column "age" ;
                 rr:language "dhhdhd"] ].

```

**Comment by sa-bpelakh** [comment]
Blank nodes only exist in context of a single result set, and are not transferrable in the manner you are using. Any serialization of a blank node (as `_:bXXXXXXX`) is completely transient. The best way to accomplish your task there is do perform a single update query, i.e.
```
PREFIX rr: <http://www.w3.org/ns/r2rml#>
DELETE DATA{
  ?om  rr:language 'dhhdhd' .
}
WHERE {
  ?om rr:language 'dhhdhd' .
}
```
Instead of a SPARQL query, you can do this programmatically as well:
```python
rr = Namespace('http://www.w3.org/ns/r2rml#')
g.remove((None, RR.language, Literal("dhhdhd")))
```
